### PR TITLE
plugins: rust

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

Port the `rust` plugin from [`snapcraft`](https://github.com/snapcore/snapcraft/blob/main/snapcraft/plugins/v2/rust.py) to `craft-parts`.

Will land this PR after https://github.com/canonical/craft-parts/pull/199.

CRAFT-955

### CODE REVIEW

File: craft_parts/plugins/rust_plugin.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    :param part_name: The part whose build environment is being validated.
+    :param env: A string containing the build step environment setup.
+    """
+
+    def validate_environment(self, *, part_dependencies: Optional[List[str]] = None):
+        """Ensure the environment has the dependencies to build Rust applications.
+
+        :param part_dependencies: A list of the parts this part depends on.
+        """
+        for dependency in ["curl", "gcc", "git"]:
```
Comment (mr-cal) [inline craft_parts/plugins/rust_plugin.py:75]: @sergiusens - I moved the dependencies from `get_build_packages()` to `validate_environment()`.  
Is this what you intended for me to do?

Comment (sergiusens) [reply craft_parts/plugins/rust_plugin.py:75]: Please leave gcc in get_build_packages and git. Completely remove curl and you should be verifying that the rustc and cargo commands are available

File: craft_parts/plugins/rust_plugin.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+        rustup_command = dedent(
+            """\
+        if ! command -v rustup 2>/dev/null; then
+            curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs |
+            sh -s -- -y --no-modify-path --profile=minimal
+            export PATH="${HOME}/.cargo/bin:${PATH}"
+        fi
+        """
+        )
```
Comment (sergiusens) [inline craft_parts/plugins/rust_plugin.py:124]: This is what the environment should validate.

**Review by sergiusens — CHANGES_REQUESTED** [review]

File: craft_parts/plugins/rust_plugin.py (hunk 1 above)
Comment (mr-cal) [reply craft_parts/plugins/rust_plugin.py]: That makes much more sense!

File: craft_parts/plugins/rust_plugin.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    def get_build_packages(self) -> Set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return {"gcc", "git"}
+
+    def get_build_environment(self) -> Dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {"PATH": "${HOME}/.cargo/bin:${PATH}"}
```
Comment (sergiusens) [inline craft_parts/plugins/rust_plugin.py:109]: this should be empty too, we don't know where folks would install their rust stuff, I guess we can leave it for now

File: tests/integration/plugins/test_rust_plugin.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
```
Comment (sergiusens) [inline tests/integration/plugins/test_rust_plugin.py:1]: can you rename this file to test_rust.py so it follows the file naming pattern in that test dir?

File: craft_parts/plugins/rust_plugin.py (hunk 3 above)
Comment (mr-cal) [reply craft_parts/plugins/rust_plugin.py:109]: Agreed on leaving it for now.
If someone creates an issue regarding a custom path for their rust binaries inside a build environment, then we can re-address.

**Review by sergiusens — APPROVED** [review]

**Pull request merged**